PDC Improvements

MrAdder opened the pull request:
Possible working fix for PDC Improvements #2 and #9

MrAdder commented on vSMR/SMRPlugin.cpp:
```diff
@@ -627,7 +649,7 @@ void CSMRPlugin::OnTimer(int Counter)
 	}
 
-	if (((clock() - timer) / CLOCKS_PER_SEC) > 10 && HoppieConnected) {
+	if (((clock() - timer) / CLOCKS_PER_SEC) > 40 && HoppieConnected) {
```
Upped mine to 100 for testing purposes and works

MrAdder commented on vSMR/SMRPlugin.cpp:
```diff
@@ -525,6 +526,27 @@ void CSMRPlugin::OnFunctionCall(int FunctionId, const char * sItemString, POINT
+		CFlightPlan FlightPlan = FlightPlanSelectASEL();
+
+		if (FlightPlan.GetFlightPlanData().GetRoute() == NULL) {
+			tmessage = "FLIGHT PLAN NOT HELD @REVERT TO VOICE PROCEDURES";
```
Feel free to change the message

AliceFord replied:
https://www.hoppie.nl/acars/system/message.html?msgid=17505280 - is an example of what vatger use - perhaps just `FLIGHT PLAN NOT HELD`?

MrAdder replied:
Done looking at now implementing it to be automatic as well as an option in the menu

MrAdder commented:
Added an Auto Connect function and the ability to turn it off and on.

Added a Retry function if the connection failed logon in use.

MrAdder commented:
**Summary of all Changes**

- Added Flight Plan Not held message in Drop down and a not tested Automated message  #2 
- Updated the Polling Rate to every 40 seconds #9 
- Added an OPT-IN Auto Connect to hoppies, and a Retry feature to attempt to connect every 2 minutes in the event of failure to login on first attempt #7 
- Fixed my Appveyor so that it works (QOL)

I believe I am ready for you to look at this now @AliceFord

MrAdder commented:
Any update on this

19wintersp commented:
Have you tested this?

MrAdder commented:
> Couldn't get the message to send for no flight plan held, but was only testing with no acft logged on, so might be why.

I can't remember if I tested this fully,


> I understood it agreed that auto-logon would not be implemented, for various reasons.

You have to opt in, it is not forced.


> I also object to re-assigning the tag function IDs without apparent reason, since this is effectively a breaking change to the public "API" (as used by ES settings files).

During my testing, it has not broken anything with the public "API"

> From my surface analysis (I can't actually compile any of this)

https://ci.appveyor.com/project/MrAdder/vsmr/build/artifacts

Build log is there aswell

19wintersp commented:
> You have to opt in, it is not forced.

I'll elucidate: from what I remember of the discussion had, connecting automatically to Hoppie's whenever logged in as a controller for a given profile (as the feature is implemented here) does not make sense, since it cannot be guaranteed that a logged-in controller on a given profile is always the appropriate controller to be logged in. Such a feature would only make sense if the client can both check that there is no existing controller logged in, and consider the precedence order against that controller for GMP functions. A function to automatically re-connect to Hoppie's if disconnected would be more sensible, but (if I'm not mistaken) Hoppie's is totally stateless and so the idea of a non-client-initiated "disconnection" doesn't really exist.

I'll leave a full code review shortly, and try to do some practical testing.

19wintersp commented on vSMR/vSMR.rc:
```diff
@@ -176,18 +176,19 @@ BEGIN
     EDITTEXT        IDC_PASSWORD,61,43,74,14,ES_PASSWORD | ES_AUTOHSCROLL
     LTEXT           "Logon Password",IDC_STATIC,7,46,52,8
     CONTROL         "Play sound on clearance request",IDC_SOUND,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,7,65,128,10
+	CONTROL         "Auto Connect and Retry", IDC_AUTO, "Button", BS_AUTOCHECKBOX | WS_TABSTOP, 7, 80, 128, 20
```
Indentation

19wintersp commented on vSMR/SMRPlugin.cpp:
```diff
@@ -615,19 +672,44 @@ void CSMRPlugin::OnTimer(int Counter)
+	}
+
+	if (!HoppieConnected && GetConnectionType() != CONNECTION_TYPE_NO && ControllerMyself().IsController() && AutoConnect == true) {
+		if (!isConnecting) {
+			// Start the connect timer
+			connectTimer = clock();
+			isConnecting = true;
+		}
+		else {
+			// Check if 10 seconds have passed since connecting
+			if (((clock() - connectTimer) / CLOCKS_PER_SEC) > 10) {
+				_beginthread(datalinkLogin, 0, NULL);
+				isConnecting = false;
+			}
+		}
```
This is a little strange. Here's my understanding of the flow as implemented, when connected as controller and with the auto-connect option enabled:

```mermaid
graph TD;
  s1[!isConnecting; connectTimer unspecified];
  s2[isConnecting; connectTimer pending];
  s3[isConnecting; connectTimer elapsed];
  s4[connection attempt spawned];
  start --> s1 --> s2 --> s2 --> s3 --> s1;
  s3 --> s4;
```

This logic means that a second connection attempt can be started if the first does not complete within the connection interval, which is undesirable. It also contains redundancy, since `isConnecting` is only ever false for a single clock cycle (ie. one second), unless the connection completes within the second, in which case it is still redundant, because it will be set back to true as soon as Hoppie's is disconnected.

This logic also interacts poorly with the reconnection logic.

19wintersp reviewed (changes requested):
I've also noticed it appears no consideration is made for whether the user chose to disconnect in the auto-connection logic (ie. if auto-connection is enabled, it will re-connect immediately if the user chooses to disconnect). I'm unsure as to whether this is intentional, but it seems undesirable (eg. for handoffs).

MrAdder commented on .gitignore:
```diff
@@ -206,3 +206,4 @@ package/
 /vSMR/euroscope_sector_providers.txt
 vSMR/vSMR.vcxproj
 /lib/include/boost_1_85_0/
+vSMR/vSMR.vcxproj
```
For some reason it iss not ignoring my local change for me to build on my PC, I can remove it when I fix the other issues that have been Identified

MrAdder commented:
I may have this fixed this week; however, I do have an interview on Tuesday that requires some reading, and work the rest of the week

19wintersp commented:
That's fine, there's no real urgency here. Since there's probably a fair deal of modification required, you may wish to consider splitting this PR into three to correspond one-to-one with the issues (as is standard) so that discussion of individual features does not block the others from being merged.
